File: src/lexer.rs
// src/lexer.rs
// Analizador léxico (scanner) implementado con Logos.
// Convierte la entrada en una secuencia de tokens tipados según la especificación.
// La salida es consumida por LALRPOP en forma de triples (start, token, end).

use logos::Logos;

/// Definición de todos los tokens del lenguaje.
///
/// Diseño:
/// - Se utiliza `#[logos(skip ...)]` para ignorar whitespace automáticamente.
/// - Orden de definición:
///     - Palabras reservadas antes que identificadores
///     - Flotantes antes que enteros
/// - Los tokens con datos (payload) usan callbacks para parsear valores.
#[derive(Logos, Debug, Clone, PartialEq)]
#[logos(skip r"[ \t\r\n]+")] // Ignora espacios, tabs y saltos de línea
pub enum Token {

    // Palabras reservadas 
    #[token("programa")]  Programa,
    #[token("inicio")]    Inicio,
    #[token("fin")]       Fin,
    #[token("nula")]      Nula,
    #[token("escribe")]   Escribe,
    #[token("mientras")]  Mientras,
    #[token("haz")]       Haz,
    #[token("si")]        Si,
    #[token("sino")]      Sino,
    #[token("vars")]      Vars,
    #[token("entero")]    Entero,
    #[token("flotante")]  Flotante,

    // Identificadores
    // (letter | _)(letter | _ | digit)*
    // Deben ir después de palabras reservadas para respetar prioridad.
    #[regex(r"[a-zA-Z_][a-zA-Z_0-9]*", |lex| lex.slice().to_string())]
    Id(String),

    // Constantes numéricas
    // Nota: flotantes antes que enteros para evitar tokenización incorrecta.
    #[regex(r"[0-9]+\.[0-9]+", |lex| lex.slice().parse::<f64>().ok())]
    CteFlot(f64),

    #[regex(r"[0-9]+", |lex| lex.slice().parse::<i64>().ok())]
    CteEnt(i64),

    // Strings
    // Formato: " ... "
    // Se eliminan las comillas en el callback.
    #[regex(r#""[^"]*""#, |lex| {
        let s = lex.slice();
        s[1..s.len()-1].to_string() 
    })]
    Letrero(String),

    // Operadores
    #[token("+")]   Plus,
    #[token("-")]   Minus,
    #[token("*")]   Star,
    #[token("/")]   Slash,
    #[token("==")]  EqEq,
    #[token("!=")]  Neq,
    #[token("<")]   Lt,
    #[token(">")]   Gt,
    #[token("=")]   Eq,

    // Puntuación
    #[token(";")]   Semi,
    #[token(",")]   Comma,
    #[token(":")]   Colon,
    #[token("(")]   LParen,
    #[token(")")]   RParen,
    #[token("{")]   LBrace,
    #[token("}")]   RBrace,
    #[token("[")]   LBracket,
    #[token("]")]   RBracket,
}

pub struct Lexer<'input> {
    inner: logos::SpannedIter<'input, Token>,
    src: &'input str,
}

impl<'input> Lexer<'input> {
    /// Crear lexer desde string fuente
    pub fn new(src: &'input str) -> Self {
        Self {
            inner: Token::lexer(src).spanned(),
            src,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct LexError {
    pub span: (usize, usize),
    pub slice: String,
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Token no reconocido: {:?} en bytes {}..{}",
            self.slice, self.span.0, self.span.1
        )
    }
}

impl<'input> Iterator for Lexer<'input> {
    type Item = Result<(usize, Token, usize), LexError>;

    fn next(&mut self) -> Option<Self::Item> {
        self.inner.next().map(|(result, span)| match result {
            Ok(tok) => Ok((span.start, tok, span.end)),

            Err(_) => {
                let slice = self.src[span.start..span.end].to_string();

                Err(LexError {
                    span: (span.start, span.end),
                    slice,
                })
            }
        })
    }
}

File: compilador/lexer.py
"""Lexer del lenguaje usando SLY."""

from sly import Lexer


class LexicalError(Exception):
    """Error léxico con el mismo formato básico del compilador."""

    def __init__(self, start: int, end: int, slice_: str):
        self.span = (start, end)
        self.slice = slice_
        super().__init__(str(self))

    def __str__(self) -> str:
        return f"Token no reconocido: {self.slice!r} en bytes {self.span[0]}..{self.span[1]}"


class PatitoLexer(Lexer):
    """Tokeniza el código fuente y conserva offsets para depuración."""

    tokens = {
        PROGRAMA,
        INICIO,
        FIN,
        NULA,
        ESCRIBE,
        MIENTRAS,
        HAZ,
        SI,
        SINO,
        VARS,
        ENTERO,
        FLOTANTE,
        REGRESA,
        ID,
        CTE_FLOT,
        CTE_ENT,
        LETRERO,
        EQEQ,
        NEQ,
        PLUS,
        MINUS,
        STAR,
        SLASH,
        LT,
        GT,
        EQ,
        SEMI,
        COMMA,
        COLON,
        LPAREN,
        RPAREN,
        LBRACE,
        RBRACE,
        LBRACKET,
        RBRACKET,
    }

    ignore = " \t\r\n"

    ID = r"[a-zA-Z_][a-zA-Z_0-9]*"
    ID["programa"] = PROGRAMA
    ID["inicio"] = INICIO
    ID["fin"] = FIN
    ID["nula"] = NULA
    ID["escribe"] = ESCRIBE
    ID["mientras"] = MIENTRAS
    ID["haz"] = HAZ
    ID["si"] = SI
    ID["sino"] = SINO
    ID["vars"] = VARS
    ID["entero"] = ENTERO
    ID["flotante"] = FLOTANTE
    ID["regresa"] = REGRESA

    CTE_FLOT = r"[0-9]+\.[0-9]+"
    CTE_ENT = r"[0-9]+"
    LETRERO = r'"[^"]*"'

    EQEQ = r"=="
    NEQ = r"!="
    PLUS = r"\+"
    MINUS = r"-"
    STAR = r"\*"
    SLASH = r"/"
    LT = r"<"
    GT = r">"
    EQ = r"="
    SEMI = r";"
    COMMA = r","
    COLON = r":"
    LPAREN = r"\("
    RPAREN = r"\)"
    LBRACE = r"\{"
    RBRACE = r"\}"
    LBRACKET = r"\["
    RBRACKET = r"\]"

    def CTE_FLOT(self, token):
        token.value = float(token.value)
        return token

    def CTE_ENT(self, token):
        token.value = int(token.value)
        return token

    def LETRERO(self, token):
        token.value = token.value[1:-1]
        return token

    def error(self, token):
        raise LexicalError(token.index, token.index + 1, token.value[0])


def iter_tokens(src: str):
    """Produce tokens con rangos para el modo `--lex`."""

    lexer = PatitoLexer()
    for token in lexer.tokenize(src):
        start = token.index
        end = start + len(str(token.value if token.type not in {"CTE_ENT", "CTE_FLOT"} else token.value))
        if token.type in {"PROGRAMA", "INICIO", "FIN", "NULA", "ESCRIBE", "MIENTRAS", "HAZ", "SI", "SINO", "VARS", "ENTERO", "FLOTANTE", "REGRESA"}:
            end = start + len(token.value)
        elif token.type == "LETRERO":
            end = start + len(token.value) + 2
        yield start, token, end
